vowpalwabbit/core/src/example_predict.cc

```diff
@@ -26,3 +26,21 @@ bool VW::example_predict::iterator::operator!=(const iterator& rhs) const { retu
 
 VW::example_predict::iterator VW::example_predict::begin() { return {feature_space.data(), indices.begin()}; }
 VW::example_predict::iterator VW::example_predict::end() { return {feature_space.data(), indices.end()}; }
+
+uint64_t VW::example_predict::calculate_feature_space_hash()
+{
+  if (feature_space_hash == 0)
```

Comment:
0 is a valid hash value which makes it impossible to tell the difference between not yet set and a hash of 0.

Reply:
this will only be zero if there are no features at all (with `--noconstant` set) e.g. `shared` example or an example with no features or features with zero weights

given that there are no features in that case a hash of `0` is fine IMO

Reply:
But 0 is a valid hash value, so std::hash could produce zero

Reply:
Since any number is a valid hash, the thing we could do is add a boolean that is set when the hash is calculated

Reply:
done